iron/gpg-packet.c

```diff
@@ -0,0 +1,1150 @@
+ *
+ *  @param infile File from which to read packet
+ *  @param key_id ID of user's public key
+ *  @param pkt Place to write PKESK packet if found
+ *  @param next_tag Tag of next packet found after last PKESK packet
+ *  @param next_len Length of next packet found after last PKESK packet
+ *  @return int 0 if successful, negative number if error
+*/
+int
+get_gpg_pkesk_packet(FILE * infile, const char * key_id, u_char * msg, gpg_tag * next_tag, int * next_len)
+{
+    int retval = IRON_ERR_NOT_FOR_USER;
+
+    *next_tag = GPG_TAG_DO_NOT_USE;
+    *next_len = -1;
+    gpg_tag tag;
+    size_t len;
+
+    if (get_gpg_tag_and_size(infile, &tag, &len) != 0) {
+        return IRON_ERR_NOT_ENCRYPTED;
+    }
+
+    while (tag == GPG_TAG_PKESK && !feof(infile)) {
+        u_char pkt_start[GPG_KEY_ID_LEN + 1];
+        if (fread(pkt_start, 1, sizeof(pkt_start), infile) == sizeof(pkt_start)) {
+            size_t left_to_read = len - (GPG_KEY_ID_LEN + 1);
+            if (*pkt_start != GPG_PKESK_VERSION) {
+                retval = IRON_ERR_NOT_ENCRYPTED;
+                break;
+            } else if (memcmp(key_id, pkt_start + 1, GPG_KEY_ID_LEN) == 0) {
+                if (fread(msg, 1, left_to_read, infile) == left_to_read) {
+                    retval = 0;
+                } else {
+                    retval = IRON_ERR_NOT_ENCRYPTED;
+                    break;
+                }
+            } else {
+                if (fseek(infile, left_to_read, SEEK_CUR) != 0) {
+                    retval = IRON_ERR_NOT_ENCRYPTED;
+                    break;
+                }
+            }
+        } else {
+            retval = IRON_ERR_NOT_ENCRYPTED;
+            break;
+        }
+
+        if (get_gpg_tag_and_size(infile, &tag, &len) != 0) {
+            retval = IRON_ERR_NOT_ENCRYPTED;
+            break;
+        }
+    }
+
+    if (retval == 0 && !feof(infile)) {
+        *next_tag = tag;
+        *next_len = len;
+    }
+
+    return retval;
+}
+
+//  The OPS packet has a fixed length - one byte each for the version, sig class, hash algo, PK
+//  algo, and last flag, and 8 bytes for the key ID.  Total of 13 bytes.
+#define ONE_PASS_SIG_LEN	13
+
+/**
+ *  Process One Pass Signature (OPS) packet
+ *
+ *  Mostly checks to make sure the fields are what we expect, but does extract the signer's key ID.
+ *
+ *  @param buf byte array from which to extract packet
+ *  @param buf_len num bytes in buf
+ *  @param key_id place to write key_id (at least GPG_KEY_ID_LEN bytes)
+ *  @return number bytes consumed from buf, or negative number if error
+ */
+int
+extract_gpg_one_pass_signature_packet(const u_char * buf, int buf_len, u_char * key_id)
+{
+	if (buf_len < ONE_PASS_SIG_LEN) return -1;
+
+	const u_char * ptr = buf;
+	if (*(ptr++) != GPG_OPS_VERSION) return -2;
+	if (*(ptr++) != GPG_SIGCLASS_BINARY_DOC) return -3;
+	if (*(ptr++) != GPG_HASHALGO_SHA256) return -4;
+	if (*(ptr++) != GPG_PKALGO_RSA_ES) return -5;
+	memcpy(key_id, ptr, GPG_KEY_ID_LEN);
```

Comment:
We really want a file full of constants that no one is expected to do anything with? They are negative numbers that are different so I could tell where it was jumping out while troubleshooting.  I could just change them all to -1.